### ISSUE TITLE
Pods in Terminating status lasts much long time when subPath is set and contains a large number of directories

### ISSUE THREAD

**Issue by Monokaix** [kubernetes/kubernetes#107667]
### What happened?

I ran a deployment with replicas 1 and use `subPathExpr` to set container's subpath dir, which uses NFS as the storage backend, and create about 100000 dirs in container's subPath dir,when delete the pod, it spent much more time to end `Terminating` status, and when the replicas increase to 10 and do the rolling update, at the same time dir nums in containers increase to 700000, it spent about 30 minutes to totally remove pod, which is intolerant.

### What did you expect to happen?

The pod is deleted fast.

### How can we reproduce it (as minimally and precisely as possible)?

1. Run a pod using subPathExpr as the volume mount.
2. Create 100000 dirs in the backend storage corresponds to the container volume mount
3. Delete the pod 
4. Inspect kubelet log：
We can find that it spent about 17s to clean the mount dir in kuelet.
```
I0118 15:28:47.192607    8551 subpath_linux.go:228] Cleaning up subpath mounts for /mnt/paas/kubernetes/kubelet/pods/c62a9d6e-8cba-44b9-b0c0-ee4d350b4f79/volume-subpaths/nfs-xxx

I0118 15:29:03.490373    8551 subpath_linux.go:313] Successfully cleaned subpath directory /mnt/paas/kubernetes/kubelet/pods/c62a9d6e-8cba-44b9-b0c0-ee4d350b4f79/volume-subpaths/nfs-xxx/test-nginx/0
```

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.10", GitCommit:"98d5dc5d36d34a7ee13368a7893dcb400ec4e566", GitTreeState:"clean", BuildDate:"2021-04-15T03:28:42Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.10-r0-22.3.0.B002", GitCommit:"189a6a4736a535c6280f69225ea3551ac2a9d013", GitTreeState:"clean", BuildDate:"2022-01-04T09:31:44Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
3.10.0-862.14.1.5.h520.eulerosv2r7.x86_64 #1 SMP
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by Monokaix** [comment]
After some debugging, I inspected the pprof with flame graph when delete a pod, it shows that `subpath.doCleanSubPaths` and `filepath.Walk` cost much more time, here is the graph:

![2](https://user-images.githubusercontent.com/16742217/150320838-637e648c-ca95-4d8b-9114-d9e217faa069.PNG)

Apparently the Walk func spent much time when doCleanPath, what we care is the top level dir,and this pr [#82698](https://github.com/kubernetes/kubernetes/pull/82698) has added some fast path to avoid much recursive time, but before gol1.16, the filepath.Walk is still less efficient, because when used in kubelet, it reads all the dir in the mount point before call the callback func in Walk even we don't use it after that, so it spends much time.

And after go 1.16, filepath.WalkDir is introduced, which is more efficient than Walk, as the comment said in golang https://github.com/golang/go/blob/master/src/io/fs/walk.go#L54
```
// The differences between WalkDirFunc compared to filepath.WalkFunc are:
//
//   - The second argument has type fs.DirEntry instead of fs.FileInfo.
//   - The function is called before reading a directory, to allow SkipDir
//     to bypass the directory read entirely.
//   - If a directory read fails, the function is called a second time
//     for that directory to report the error.
//
```

WalkDirFunc is called before reading a directory, so it can save much time when there are lots of dirs in the first level of mount points. So it's time wasting and meaningless when we use Walk as so far because we traverse the dir but we didn't use it and return SkipDir err, the code is [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/subpath/subpath_linux.go#L275)

After I replace it with `filepath.WalkDir`, the doCleanSubPaths takes only about 41ms, which is a huge boost than 17s before.

**Comment by Monokaix** [comment]
I tried to fix this, and pr is here[#107669](https://github.com/kubernetes/kubernetes/pull/107669)
